src/Framework/Framework/Compilation/ControlTree/ControlTreeResolverBase.cs

```diff
@@ -699,7 +695,12 @@ protected virtual void ResolveControlContentImmediately(IAbstractControl control
             foreach (var node in content)
             {
                 var child = ProcessNode(control, node, control.Metadata, dataContext);
-                if (child != null) treeBuilder.AddChildControl(control, child);
+                if (child != null)
+                {
+                    treeBuilder.AddChildControl(control, child);
+                    if (!child.Metadata.Type.IsAssignableTo(ResolvedTypeDescriptor.Create(typeof(DotvvmControl))))
+                        node.AddError($"Content control must inherit from DotvvmControl, but {child.Metadata.Type} doesn't.");
+                }
```

Comment:
This has to be here, because this case was uncovered and one test tested that the C# compilation failed :facepalm:  So this fixes the underlying problem